torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
+    with ctx.redispatch_to_next():
+        unwrapped_outputs = triton_kernel_wrapper_functional(
+            kernel=kernel, grid=grid, kwargs=unwrapped_kwargs
+        )
+
+    assert unwrapped_outputs.keys() == kwargs.keys()
+    for key, output_arg in unwrapped_outputs.items():
+        if not isinstance(output_arg, Tensor):
+            continue
+        input_arg = kwargs[key]
+        assert isinstance(input_arg, Tensor)
+
+        ctx.replace(input_arg, output_arg)
+        ctx.commit_update(input_arg)
+        ctx.sync(input_arg)
+    return None
+
+
+@triton_kernel_wrapper_functional.py_impl(DispatchKey.CompositeExplicitAutograd)
+def triton_kernel_wrapper_functional_dense(*, kernel, grid, kwargs):
+    kwargs = {
+        key: (clone_preserve_strides(val) if isinstance(val, Tensor) else val)
+        for key, val in kwargs.items()
+    }
+    triton_kernel_wrapper_mutation(kernel=kernel, grid=grid, kwargs=kwargs)
+    return kwargs
+
+
+@triton_kernel_wrapper_functional.py_impl(FakeTensorMode)
+def triton_kernel_wrapper_functional_fake_tensor_mode(mode, *, kernel, grid, kwargs):
+    with mode:
+        return {
             key: (clone_preserve_strides(val) if isinstance(val, Tensor) else val)
             for key, val in kwargs.items()
         }
-        triton_kernel_wrapper_mutation(kernel=kernel, grid=grid, kwargs=kwargs)
-        return kwargs
 
 
-triton_kernel_wrapper_functional = TritonKernelWrapperFunctional()
+@triton_kernel_wrapper_functional.py_impl(ProxyTorchDispatchMode)
+def triton_kernel_wrapper_functional_proxy_torch_dispatch_mode(
+    mode, *, kernel, grid, kwargs
+):
+    if mode.enable_tracing:
+        return trace_triton_kernel_wrapper(
+            mode,
+            triton_kernel_wrapper_functional,
+            kernel=kernel,
+            grid=grid,
+            kwargs=kwargs,
+        )
+    else:
+        return triton_kernel_wrapper_functional(kernel=kernel, grid=grid, kwargs=kwargs)
+
+
+@triton_kernel_wrapper_functional.py_functionalize_impl
+def triton_kernel_wrapper_functional_functionalize(ctx, kernel, grid, kwargs):
+    unwrapped_kwargs = ctx.unwrap_tensors(kwargs)
+    with ctx.redispatch_to_next():
+        outputs = triton_kernel_wrapper_functional(
+            kernel=kernel, grid=grid, kwargs=unwrapped_kwargs
+        )
+        return ctx.wrap_tensors(outputs)
+
+
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.PythonDispatcher)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.PythonTLSSnapshot)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.ADInplaceOrView)
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.BackendSelect)
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutocastCPU)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutocastCUDA)  # type: ignore[attr-defined]
+triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutogradCUDA)
```

Comment:
What happens when we encounter a CPU Tensor (and try to pass it to this triton kernel underneath torch.compile?)

We might need to update this to `triton_kernel_wrapper_mutation.fallthrough(DispatchKey.Autograd)`, if that works. Otherwise we should probably also add a `triton_kernel_wrapper_mutation.fallthrough(DispatchKey.AutogradCPU)`.